addons/sale_project/models/sale_order.py

```diff
@@ -88,56 +105,88 @@ def _action_confirm(self):
-        else:
-            action = self.env["ir.actions.actions"]._for_xml_id("project.action_view_task")
-            action['context'] = {}  # erase default context to avoid default filter
-            if len(self.tasks_ids) > 1:  # cross project kanban task
-                action['views'] = [[False, 'kanban'], [list_view_id, 'tree'], [form_view_id, 'form'], [False, 'graph'], [False, 'calendar'], [False, 'pivot']]
-            elif len(self.tasks_ids) == 1:  # single task -> form view
-                action['views'] = [(form_view_id, 'form')]
-                action['res_id'] = self.tasks_ids.id
+        default_line = next(sol for sol in self.order_line if sol.product_id.detailed_type == 'service')
+
+        action = self.env["ir.actions.actions"]._for_xml_id("project.action_view_task")
+        action['context'] = {}  # erase default context to avoid default filter
+        if self.tasks_count > 1:  # cross project kanban task
+            action['views'] = [[kanban_view_id, 'kanban'], [list_view_id, 'tree'], [form_view_id, 'form'], [False, 'graph'], [False, 'calendar'], [False, 'pivot']]
+        else:  # 1 or 0 tasks -> form view
+            action['views'] = [(form_view_id, 'form')]
+            action['res_id'] = self.tasks_ids.id
         # filter on the task of the current SO
         action.setdefault('context', {})
-        action['context'].update({'search_default_sale_order_id': self.id})
+        # set default project
+        default_project_id = default_line.project_id.id or self.project_id.id or (len(self.project_ids) > 0 and self.project_ids[0].id)
+
+        action['context'].update({
+            'search_default_sale_order_id': self.id,
+            'default_sale_order_id': self.id,
+            'default_sale_line_id': default_line.id,
+            'default_partner_id': self.partner_id.id,
+            'default_project_id': default_project_id,
+            'default_user_ids': [self.env.uid],
+        })
+        return action
+
+    def action_create_project(self):
+        self.ensure_one()
+        if not self.order_line:
+            return {'type': 'ir.actions.act_window_close'}
+
+        sorted_line = self.order_line.sorted(key=lambda sol: sol.sequence)
+        default_sale_line = next(sol for sol in sorted_line if sol.product_id.detailed_type == 'service')
+        action = self.env["ir.actions.actions"]._for_xml_id("project.open_create_project")
+        action['context'] = {
+            'default_sale_order_id': self.id,
+            'default_sale_line_id': default_sale_line.id,
+            'default_partner_id': self.partner_id.id,
+            'default_user_ids': [self.env.uid],
+            'default_allow_billable': 1,
+            'hide_allow_billable': True,
+            'default_company_id': self.company_id.id
+        }
+        if default_sale_line.product_id.service_policy == 'delivered_milestones':
+            action['context'].update({'generate_milestone': True})
         return action
```

Comment:
```suggestion
        return {
            **self.env["ir.actions.actions"]._for_xml_id("project.open_create_project"),
            'context': {
                'default_sale_order_id': self.id,
                'default_sale_line_id': default_sale_line.id,
                'default_partner_id': self.partner_id.id,
                'default_user_ids': [self.env.uid],
                'default_allow_billable': 1,
                'hide_allow_billable': True,
                'default_company_id': self.company_id.id
                'generate_milestone': default_sale_line.product_id.service_policy == 'delivered_milestones',
            },
        }
```
No? :slightly_smiling_face:
If not, no need to 'update' for one key. Just `dict[key] = value`.